### PR TITLE
Jamie_dev

### PR DESCRIPTION
**Pull request by jameson-c**
V 1.0.0 of read / save package. 

pip install git+https://github.com/UI-Research/RreadFEH@jamie_dev will get feh_io.

Formatted in SRC layout which could be altered. This is because I found during testing that setup.py is deprecated and TOML files are now used to package things instead. https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/ for details.

### CODE REVIEW

File: Python/test-feh.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+"""
+Test for DYNASIM FEH package.
+
+This script tests the DYNASIM feh_io module.
+"""
+
+from feh_io.read_feh import read_feh_data_file, read_parquet_2
```
Comment (dc0sic) [inline Python/test-feh.py]: This is fine as is, but I just want to clarify what I meant when I ask you for a different kind of import. I had in mind something like this:
```
import feh_io
feh_io.read_feh_data_file
```
Or, with using `from`:
```
from feh_io import read_feh_data_file
```
I am not sure how to do that, but I think there is a way to specify names that a module exports in its `__init__.py`. This is definitely not a big deal, we can live with this.

File: Python/test-feh.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
+"""
```
Comment (dc0sic) [inline Python/test-feh.py:1]: I am a fan of unit tests. Have you considered using Pytest?

**Review by dc0sic — APPROVED** [review]
This looks great, thank you. I left a couple of comments but no show stoppers. 

One thing I realized is that `read_feh_data_file` does not implement chunked reading! It implements partial reading that allows the caller to specify the number of observations to read, but it does not retain the file position or allow the caller to specify it, which would be required for reading multiple chunks. I think that would be a good next task.

File: Python/test-feh.py (hunk 1 above)
Comment (jameson-c) [reply Python/test-feh.py]: ok so this was accomplished with the prior configuration where the __init__.py combined all of the functions in read_feh and save_feh together. I don't think there's another way to do it unless you want the save module and the read module to be combined into one. Thoughts?

File: Python/test-feh.py (hunk 2 above)
Comment (jameson-c) [reply Python/test-feh.py:1]: I am basically doing some things similar to this but without pytest, thanks for the recommendation! the next go-around I will try and convert some of these tests.

File: Python/test-feh.py (hunk 1 above)
Comment (dc0sic) [reply Python/test-feh.py]: I am not sure what "prior configuration" you're referring to. It seems that what I have in mind could be accomplished by import statements in `__init__.py`:
```
from .read_feh import read_feh_data_file
from .save_feh import save_feh_parquet
```
See item 2 in https://coderslegacy.com/python-init-py-best-practices/.

Comment (jameson-c) [reply Python/test-feh.py]: ah sweet! I can implement this :) Thanks for pointing it out.

Comment (jameson-c) [reply Python/test-feh.py]: addressed and tested!

Comment (dc0sic) [reply Python/test-feh.py]: Thank you!

**Comment by jameson-c** [comment]
> This looks great, thank you. I left a couple of comments but no show stoppers.
> 
> One thing I realized is that `read_feh_data_file` does not implement chunked reading! It implements partial reading that allows the caller to specify the number of observations to read, but it does not retain the file position or allow the caller to specify it, which would be required for reading multiple chunks. I think that would be a good next task.

In reference to this, Damir, I have poked around a bit today. It looks like in order to use the [offset ](https://numpy.org/doc/stable/reference/generated/numpy.fromfile.html) argument of fromfile (the currently used method) you need to know the amount of _bytes_ to offset. It appears that one solution would be to 
1. read the first chunk in for `count = n` **without** an offset. 
2. get [itemsize](https://numpy.org/doc/stable/glossary.html#term-itemsize) of that chunk. 
3. deduce itemsize per record for `count = n` using `data.dtype.itemsize`
4. Set `offset` in `np.fromfile()` to this value multiplied by the count of values for ensuing records.
5. Increase the offset size with each iteration of chunking.
Not sure if this tracks with what you were imagining. Thoughts?

**Comment by dc0sic** [comment]
> > One thing I realized is that `read_feh_data_file` does not implement chunked reading! It implements partial reading that allows the caller to specify the number of observations to read, but it does not retain the file position or allow the caller to specify it, which would be required for reading multiple chunks. I think that would be a good next task.
> 
> In reference to this, Damir, I have poked around a bit today. It looks like in order to use the [offset ](https://numpy.org/doc/stable/reference/generated/numpy.fromfile.html) argument of fromfile (the currently used method) you need to know the amount of _bytes_ to offset. It appears that one solution would be to
> 
> 1. read the first chunk in for `count = n` **without** an offset.
> 2. get [itemsize](https://numpy.org/doc/stable/glossary.html#term-itemsize) of that chunk.
> 3. deduce itemsize per record for `count = n` using `data.dtype.itemsize`
> 4. Set `offset` in `np.fromfile()` to this value multiplied by the count of values for ensuing records.
> 5. Increase the offset size with each iteration of chunking.
>    Not sure if this tracks with what you were imagining. Thoughts?

Yes, that makes sense. I did implement [chunked reading in the R implementation](https://github.com/UI-Research/RreadFEH/blob/67f0b6dbc2b51668c17bf6c4e060811701d404a2/R/read_feh.R#L406), although only internally, inside the reader function. This was to allow selecting columns on chunks and avoid having to read the entire file before being able to do the select. You can see the mechanics of how I dealt with the calculation of offset.

However, what we want is to allow the caller to read chunks by calling the reader function multiple times, which means that we need to keep track of the offset between the calls. Perhaps we should use a class?

**Comment by jameson-c** [comment]
So ideally, it appears that we could do something to specify a class of data parametrizations from the YAML config (assuming that is the route we go) that takes:

**Comment by jameson-c** [comment]
So ideally, it appears that we could do something to specify a class of data parametrizations from the YAML config (assuming that is the route we go) that contains:
- minimum chunk size (last chunk will be variable in size)
- variables used in the analysis 
- data for a chunk pulled after initializing the chunk class
- chunk offset contributed by this chunk
- cumulative chunk offset contributed by prior chunks
This could be instantiated when initializing the read function. Additionally, we could submit an argument to describe whether just 1 chunk of the given size should be taken (for testing purposes). Thoughts?

**Comment by dc0sic** [comment]
We will want to have a YAML config but that will be part of the model. In this package, I think it's best to keep it simple and pass these parameters to functions or methods, depending on what route we want to take. 

It seems to me that having a class would be helpful for maintaining a file offset and a file object. Something like this: 
```
class FehReader:
    def __init__(self, filename, chunksize=-1, varlist=None):
        self.bytes_read=0
        self.file = open(filename)
        self.file_size = os.path.getsize(filename)
    def read(self):
        data = self.file.read(chunksize)
        self.bytes_read = self.bytes_read + len(data)
        if self.bytes_read >= self.file_size:
            self.file(close)
        return data
```
When reading in chunks, the caller would call `read` from a loop. Error handling and detecting the end of the file by the caller would need to be added.

**Pull request merged**